panther_lights/src/animation/image_animation.cpp

```diff
@@ -0,0 +1,169 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <panther_lights/animation/image_animation.hpp>
+
+#include <algorithm>
+#include <cstdint>
+#include <filesystem>
+#include <regex>
+#include <stdexcept>
+#include <string>
+
+#include <yaml-cpp/yaml.h>
+
+#include <ament_index_cpp/get_package_prefix.hpp>
+#include <ament_index_cpp/get_package_share_directory.hpp>
+#include <boost/gil.hpp>
+#include <boost/gil/extension/io/png.hpp>
+#include <boost/gil/extension/numeric/resample.hpp>
+#include <boost/gil/extension/numeric/sampler.hpp>
+
+namespace panther_lights
+{
+
+void ImageAnimation::Initialize(
+  const YAML::Node & animation_description, const std::size_t num_led,
+  const float controller_frequency)
+{
+  Animation::Initialize(animation_description, num_led, controller_frequency);
+
+  if (!animation_description["image"]) {
+    throw std::runtime_error("No 'image' in animation description");
+  }
+
+  const auto image_path = ParseImagePath(animation_description["image"].as<std::string>());
+  gil::rgba8_image_t base_image;
+  gil::read_and_convert_image(std::string(image_path), base_image, gil::png_tag());
+  image_ = RGBAImageResize(base_image, this->GetNumberOfLeds(), this->GetAnimationLength());
+
+  if (animation_description["color"]) {
+    RGBAImageConvertColor(image_, animation_description["color"].as<std::uint32_t>());
+  }
+}
+
+std::vector<std::uint8_t> ImageAnimation::UpdateFrame()
+{
+  std::vector<std::uint8_t> frame;
+  for (std::size_t i = 0; i < this->GetNumberOfLeds(); i++) {
+    auto pixel = gil::const_view(image_)(i, this->GetAnimationIteration());
+    frame.push_back(pixel[0]);
+    frame.push_back(pixel[1]);
+    frame.push_back(pixel[2]);
+    frame.push_back(pixel[3]);
+  }
+
+  return frame;
+}
+
+std::filesystem::path ImageAnimation::ParseImagePath(const std::string & image_path) const
+{
+  std::filesystem::path global_img_path;
+  if (!std::filesystem::path(image_path).is_absolute()) {
+    if (image_path[0] == '$') {
+      std::smatch match;
+      if (std::regex_search(image_path, match, std::regex("^\\$\\(find .*\\)"))) {
+        std::string ros_pkg_expr = match[0];
+        std::string ros_pkg = std::regex_replace(
+          ros_pkg_expr, std::regex("^\\$\\(find \\s*|\\)$"), "");
+
+        auto img_relative_path = image_path;
+        img_relative_path.erase(0, ros_pkg_expr.length());
+
+        try {
+          global_img_path = ament_index_cpp::get_package_share_directory(ros_pkg) +
+                            img_relative_path;
+        } catch (const ament_index_cpp::PackageNotFoundError & e) {
+          throw std::runtime_error("Can't find ROS package: " + ros_pkg);
+        }
+
+      } else {
+        throw std::runtime_error("Can't process substitution expression");
+      }
+    } else {
+      throw std::runtime_error("Invalid image path");
+    }
+  } else {
+    global_img_path = image_path;
+  }
```

Comment:
```suggestion
  if (std::filesystem::path(image_path).is_absolute()) {
    return std::filesystem::path(image_path);
  }

  if (image_path[0] != '$') {
    throw std::runtime_error("Invalid image path");
  }

  std::smatch match;
  if (!std::regex_search(image_path, match, std::regex("^\\$\\(find .*\\)"))) {
    throw std::runtime_error("Can't process substitution expression");
  }

  std::string ros_pkg_expr = match[0];
  std::string ros_pkg = std::regex_replace(ros_pkg_expr, std::regex("^\\$\\(find \\s*|\\)$"), "");
  std::string img_relative_path = image_path.substr(ros_pkg_expr.length());

  try {
    return ament_index_cpp::get_package_share_directory(ros_pkg) + img_relative_path;
  } catch (const ament_index_cpp::PackageNotFoundError &e) {
    throw std::runtime_error("Can't find ROS package: " + ros_pkg);
  }
```
It's only suggestion. Please review it thoroughly. :p